### PR TITLE
Feature save transaction

### PR DESCRIPTION
**Pull request by jubalm**
- [x] switch transaction page view based on transfer type
- [x] saved transactions hook
- [x] saved transactions sidebar on transfer and transactions page
- [ ] dialog for save transaction

| favorites 2 |
|:--:|
| *Saved Transfer Sidebar* |

| Screenshot 2023-06-13 at 12 39 29 PM |
|:--:|
| *Save transfer Form* |

| Screenshot 2023-06-13 at 12 39 12 PM |
|:--:|
| *Save transfer Complete* |

### CODE REVIEW

**Comment by MicahZoltu** [comment]
For the text in the box, can we have it say something like this depending on which fields are saved and have data?
```
5
ETH
0xabcd...
5 ETH
5 to 0xabcd...
5 ETH to 0xabcd...
ETH to 0xabcd...
```

File: app/ts/components/Favorites.tsx [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+type PromoteButtonProps = {
+	show: boolean
+	favorite: FavoriteModel
+	index: number
+}
+
+const MoveUpButton = (props: PromoteButtonProps) => {
+	const { swapIndex } = useFavorities()
```
Comment (KillariDev) [inline app/ts/components/Favorites.tsx:58]:
```suggestion
	const { swapIndex } = useFavorite()
```

File: app/ts/components/TransactionPage/SaveTransfer.tsx [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+		onSubmit()
+	}
+
+	return (
+		<div class='my-4'>
+			<div class='font-bold text-2xl mb-2'>Save Transfer</div>
+			<div class='border border-dashed border-white/30 p-4'>
+				<div class='flex flex-col md:flex-row-reverse items-center gap-4'>
+					<div class='shrink w-full'>
+						<p class='text-white/50 text-sm'>Prevent accidental inputs by saving this transfer so you can quickly do this again later. Add a label to this transfer and hit save to continue.</p>
```
Comment (KillariDev) [inline app/ts/components/TransactionPage/SaveTransfer.tsx:52]: I feel the label should be optional

File: app/ts/components/TransactionPage/SaveTransfer.tsx [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+					<svg width='60' height='60' viewBox='0 0 24 24' xmlns='http://www.w3.org/2000/svg'>
+						<g fill='none' fillRule='evenodd'>
+							<path d='M0 0h24v24H0z' />
+							<circle stroke='currentColor' strokeWidth='2' strokeLinecap='round' cx='12' cy='12' r='9' />
+							<path d='m8.5 12.5 1.651 2.064a.5.5 0 0 0 .744.041L15.5 10' stroke='currentColor' strokeWidth='2' strokeLinecap='round' />
+						</g>
+					</svg>
+					<div>
+						<div class='font-bold text-lg'>Transfer Saved!</div>
+						<div class='text-white/50 leading-tight'>This transfer was added to the sidebar so you can use it as a starting point the next time you transfer.</div>
```
Comment (KillariDev) [inline app/ts/components/TransactionPage/SaveTransfer.tsx:82]:
```suggestion
						<div class='text-white/50 leading-tight'>This transfer was added to the sidebar so you can use it as a starting point for the next time you transfer.</div>
```

File: app/ts/store/favorites.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	amount: string
+}
+
+const FAVORITES_CACHE_ID = 'favorites'
+
+const getCachedFavorites = () => {
+	const cached = localStorage.getItem(FAVORITES_CACHE_ID)
+	if (cached === null) return []
+	try {
+		const parsed = JSON.parse(cached) as unknown
```
Comment (KillariDev) [inline app/ts/store/favorites.ts:19]: should use funtypes

File: app/ts/store/favorites.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+export function isFavorite(data: unknown): data is FavoriteModel {
+	return data !== null && typeof data === 'object' && 'label' in data && typeof data.label === 'string' && 'source' in data && typeof data.source === 'string' && 'recipientAddress' in data && typeof data.recipientAddress === 'string'
+}
+
+const favorites = signal<FavoriteModel[]>(getCachedFavorites())
+
+effect(() => {
+	const uniqueTxns = Array.from(new Set(favorites.value))
+	localStorage.setItem(FAVORITES_CACHE_ID, JSON.stringify(uniqueTxns))
```
Comment (KillariDev) [inline app/ts/store/favorites.ts:38]: should use funtypes

File: app/ts/store/favorites.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+})
+
+export function useFavorities() {
+	const syncCacheChange = (event: StorageEvent) => {
+		const newValue = event.newValue !== null ? (JSON.parse(event.newValue) as FavoriteModel[]) : []
```
Comment (KillariDev) [inline app/ts/store/favorites.ts:43]: should use funtypes

File: app/ts/store/transaction.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+	const providers = useProviders()
+	const { value: transactionResponse, waitFor: waitForResponse, reset: resetResponse } = useAsyncState<TransactionResponse>()
+	const { value: transactionReceipt, waitFor: waitForReceipt, reset: resetReceipt } = useAsyncState<TransactionReceipt>()
+
+	const getTransactionResponse = (transactionHash: string) => {
+		waitForResponse(async () => {
+			const provider = providers.getbrowserProvider()
+			const result = await provider.getTransaction(transactionHash)
+			// TransactionResult can actually be null?
+			if (result === null) throw new Error('Transaction was not found in chain!')
```
Comment (KillariDev) [inline app/ts/store/transaction.ts:17]:
```suggestion
			if (result === null) throw new Error('Transaction was not found on chain!')
```

**Review by KillariDev — APPROVED** [review]

File: app/ts/components/TransactionPage/SaveTransfer.tsx (hunk 2 above)
Comment (jubalm) [reply app/ts/components/TransactionPage/SaveTransfer.tsx:52]: Let me track as discussion for now in #145

File: app/ts/store/favorites.ts (hunk 5 above)
Comment (jubalm) [reply app/ts/store/favorites.ts:38]: Will redo in #121

File: app/ts/store/favorites.ts (hunk 4 above)
Comment (jubalm) [reply app/ts/store/favorites.ts:19]: Will redo in #121

File: app/ts/store/favorites.ts (hunk 6 above)
Comment (jubalm) [reply app/ts/store/favorites.ts:43]: Will redo in #121 all JSON parsing

**Comment by jubalm** [comment]
> For the text in the box, can we have it say something like this depending on which fields are saved and have data?
> 
> ```
> 5
> ETH
> 0xabcd...
> 5 ETH
> 5 to 0xabcd...
> 5 ETH to 0xabcd...
> ETH to 0xabcd...
> ```

Screenshot 2023-06-15 at 8 57 35 AM

**Pull request merged**